spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowServerConfiguration.java

```diff
@@ -161,10 +175,97 @@ public void configurePathMatch(PathMatchConfigurer configurer) {
 	}
 
 	@Bean
-	@ConditionalOnMissingBean
+	@ConditionalOnBean(Server.class)
+	public TaskExplorer taskExplorer(DataSource dataSource, Server server) {
+		return new SimpleTaskExplorer(new TaskExecutionDaoFactoryBean(dataSource));
+	}
+
+	@Bean
+	@ConditionalOnMissingBean(Server.class)
 	public TaskExplorer taskExplorer(DataSource dataSource) {
-		JdbcTaskExplorerFactoryBean factoryBean = new JdbcTaskExplorerFactoryBean(dataSource);
-		return factoryBean.getObject();
+		return new SimpleTaskExplorer(new TaskExecutionDaoFactoryBean(dataSource));
+	}
+
+	@Bean
+	public TaskJobRepository taskJobExecutionRepository(JobService service, TaskExplorer taskExplorer){
+		return new TaskJobRepository(service, taskExplorer);
+	}
+
+	@Bean
+	public SimpleJobServiceFactoryBean simpleJobServiceFactoryBean(DataSource dataSource,
+			JobRepositoryFactoryBean repositoryFactoryBean) {
+		SimpleJobServiceFactoryBean factoryBean = new SimpleJobServiceFactoryBean();
+		factoryBean.setDataSource(dataSource);
+		try {
+			factoryBean.setJobRepository(repositoryFactoryBean.getObject());
+			factoryBean.setJobLocator(new MapJobRegistry());
+			factoryBean.setJobLauncher(new SimpleJobLauncher());
+			factoryBean.setDataSource(dataSource);
+		}
+		catch (Exception e) {
+			throw new IllegalStateException(e);
+		}
+		return factoryBean;
+	}
+
+	@Bean
+	@ConditionalOnBean(Server.class)
+	public JobRepositoryFactoryBean jobRepositoryFactoryBeanForServer(DataSource dataSource, Server server, DataSourceTransactionManager dataSourceTransactionManager){
```

Comment:
Good question.  In short when using the embedded  h2 database, all beans that require the dataSource have to wait for the H2Server to come up so that they  can have access to the h2  url and port (even though the datasource is now available). So these beans have to wait for the server.   But when using another database no H2 server will be started, and the other set of conditional beans handle that scenario.